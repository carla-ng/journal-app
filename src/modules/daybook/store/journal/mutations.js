
// export const myMutation = ( state ) => {

//import state from "./state"

// }


export const setEntries = ( state, entries ) => {
    state.entries = [ ...state.entries, ...entries ]
    state.isLoading = false
}


export const updateEntry = ( state, entry ) => {
    const index = state.entries.map( e => e.id ).indexOf( entry.id )
    state.entries[index] = entry
}


export const addEntry = ( state, entry ) => {
    state.entries = [ entry, ...state.entries ]

}