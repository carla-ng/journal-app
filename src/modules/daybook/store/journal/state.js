
export default () => ({
    isLoading: true,
    entries: [
        {
            id: '1',
            date: new Date().toDateString(),
            text: 'Lorem exercitation labore ipsum duis mollit exercitation veniam. Dolor in dolor exercitation consectetur consequat ex duis pariatur sunt. Labore amet amet nulla occaecat ad sint. Aliqua aliquip elit laboris reprehenderit veniam.',
            picture: null,
        },
        {
            id: '2',
            date: new Date().toDateString(),
            text: 'Qui minim tempor est consectetur ullamco cillum magna cillum reprehenderit non nostrud fugiat adipisicing nostrud. Consequat mollit mollit veniam quis nisi eu nisi Lorem tempor. Eu eiusmod non cupidatat adipisicing consequat occaecat mollit labore. Enim ut magna reprehenderit enim ea culpa ullamco ullamco nisi.',
            picture: null,
        },
        {
            id: '3',
            date: new Date().toDateString(),
            text: 'Commodo excepteur in et veniam pariatur ad Lorem in. Cillum consectetur nostrud irure exercitation id. Aliquip ad proident deserunt excepteur nulla cupidatat ut aliquip. Sint ex dolor do anim.',
            picture: null,
        }
    ]
})