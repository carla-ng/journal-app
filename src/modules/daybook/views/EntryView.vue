<template>
  
  <template v-if="entry">
    <div class="entry-title d-flex justify-content-between p-2">
      <div>
          <span class="text-success fs-3 fw-bold">{{ day }}</span>
          <span class="mx-1 fs-3">{{ month }}</span>
          <span class="mx-2 fs-4 fw-light">{{ yearAndDay }}</span>
      </div>

      <div>
        <button class="btn btn-danger mx-2">
          Delete
          <i class="fa fa-trash-alt"></i>
        </button>

        <button class="btn btn-primary">
          Upload picture
          <i class="fa fa-upload"></i>
        </button>
      </div>
    </div>

    <hr>

    <div class="d-flex flex-column px-3 h-75">
      <textarea v-model="entry.text" placeholder="What happened today?"></textarea>
    </div>

    <img
        src="https://images.unsplash.com/photo-1605979257913-1704eb7b6246?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1470&q=80"
        alt="entry-picture"

        class="img-thumbnail">
  </template>

  <Fab icon="fa-save" />
  
</template>


<script>
import { defineAsyncComponent } from 'vue'
import { mapGetters } from 'vuex'
import getDayMonthYear from '../helpers/getDayMonthYear'

export default {
    props: {
      id: {
        type: String,
        required: true
      }
    },
    components: {
      Fab: defineAsyncComponent( () => import('../components/Fab.vue') )
    },
    data() {
      return {
        entry: null
      }
    },
    computed: {
      ...mapGetters('journal', ['getEntryById']),
      day() {
        const { day } = getDayMonthYear( this.entry.date )
        return day
      },
      month() {
        const { month } = getDayMonthYear( this.entry.date )
        return month
      },
      yearAndDay() {
        const { yearAndDay } = getDayMonthYear( this.entry.date )
        return yearAndDay
      }
    },
    methods: {
      loadEntry() {
        const entry = this.getEntryById( this.id )

        if ( !entry ) return this.$router.push({ name: 'no-entry' })
        this.entry = entry
      }
    },
    created() {
      this.loadEntry()
    },
    watch: {
      id() {
        this.loadEntry()
      }
    }
}
</script>

<style lang="scss" scoped>
textarea {
  border: none;
  font-size: 20px;
  height: 100%;

  &:focus {
    outline: none;
  }
}

img {
  position: fixed;
  bottom: 150px;
  right: 20px;

  box-shadow: 0 5px 10px rgba($color: #000, $alpha: 0.2);
  width: 200px;
}
</style>