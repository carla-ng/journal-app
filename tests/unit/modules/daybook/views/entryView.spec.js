import EntryView from '@/modules/daybook/views/EntryView.vue'
import journal from '@/modules/daybook/store/journal'
import Swal from 'sweetalert2'

import { shallowMount } from '@vue/test-utils'
import { createStore } from 'vuex'

import { journalState } from '../../../mock-data/test-journal-state'


const createVuexStore = ( initialState ) => createStore({
    modules: {
        journal: {
            ...journal,
            state: { ...initialState }
        }
    }
})

jest.mock('sweetalert2', () => ({
    fire: jest.fn(),
    showLoading: jest.fn(),
    close: jest.fn()
}) )


describe('EntryView tests', () => {

    const store = createVuexStore( journalState )
    store.dispatch = jest.fn()

    const mockRouter = {
        push: jest.fn()
    }

    let wrapper

    beforeEach(() => {
        jest.clearAllMocks()
        wrapper = shallowMount( EntryView, {
            props: {
                id: 'XYZ456'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [ store ]
            }
        })
    })


    test('Must redirect user when ID does not exist', () => {
        
        const wrapper = shallowMount( EntryView, {
            props: {
                id: 'This ID does not exist in store'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [ store ]
            }
        })

        expect( mockRouter.push ).toHaveBeenCalledWith({ name: 'no-entry' })
     })

     test('Must show entry correctly', () => {
        expect( wrapper.html() ).toMatchSnapshot()
        expect( mockRouter.push ).not.toHaveBeenCalled()
     })

     test('Must delete entry and leave its page', (done) => {
        Swal.fire.mockReturnValueOnce( Promise.resolve({ isConfirmed: true }) )
        wrapper.find('.btn-danger').trigger('click')
        
        expect( Swal.fire ).toHaveBeenCalledWith({
            title: 'Are you sure you want to delete this entry?',
            text: 'Once it is deleted you cannot get it back',
            showDenyButton: true,
            confirmButtonText: 'Yes, I am sure'
        })

        setTimeout(() => {
            expect( store.dispatch ).toHaveBeenCalledWith('journal/deleteEntry', 'XYZ456')
            expect( mockRouter.push ).toHaveBeenCalled()
            done()
        }, 1);
     })

})