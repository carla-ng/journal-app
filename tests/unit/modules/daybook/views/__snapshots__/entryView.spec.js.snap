// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntryView tests Must show entry correctly 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">16</span><span class="mx-1 fs-3">July</span><span class="mx-2 fs-4 fw-light">2021, Friday</span></div>
  <div><input type="file" accept="image/*" style="display: none;"><button class="btn btn-danger mx-2"> Delete <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary"> Upload picture <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="What happened today?"></textarea></div>
<img src="https://res.cloudinary.com/dkfdkhhsd/image/upload/v1650664017/fxectjbnv5k5yc0v5eq6.jpg" alt="entry-picture" class="img-thumbnail">
<!--v-if-->
<fab-stub icon="fa-save"></fab-stub>
`;
